src/vm/dllimport.cpp

```diff
@@ -7076,6 +7076,15 @@ EXTERN_C LPVOID STDCALL NDirectImportWorker(NDirectMethodDesc* pMD)
             _ASSERTE(pThread->GetFrame()->GetVTablePtr() == InlinedCallFrame::GetMethodFrameVPtr());
 
             CONSISTENCY_CHECK(pMD->IsNDirect());
+
+            if (pMD->GetModule()->IsReadyToRun())
+            {
+                // We need pMD in the case of an exception unwinding, so we can test if the pinvoke method is
+                // in a R2R module, and if so, properly pop the PInvoke frame from the thread. This is because
+                // the JIT_PInvokeEnd helper that pops the frame will not run when there's an exception.
+                ((InlinedCallFrame*)pThread->GetFrame())->m_Datum = pMD;
```

Comment:
What will happen if the exception is thrown from the method in steady state? Note that we do support exception interop on Windows.

Reply:
I'm not sure if I understood this correctly, but if you're referring to exceptions being thrown by the method with pinvoke away from any pinvoke code, then this works (also updated the test with this scenario)

Reply:
Or did you mean after a tiered re-compilation of the method?

Reply:
The case I am wondering about is when the `DllImport` method throws exception in C/C++. I do not see a test for that in your changes.

To add coverage for this, you can add `[DllImport]` for `RaiseException` that will get called in a loop.

Reply:
Ok, that one works. I added it to the test